reconcile/utils/differ.py

```diff
@@ -0,0 +1,215 @@
+            identical[k] = diff_pair
+        else:
+            change[k] = diff_pair
+    return DiffResult(
+        add=add,
+        delete=delete,
+        change=change,
+        identical=identical,
+    )
+
+
+def diff_any_iterables(
+    current: Iterable[Current],
+    desired: Iterable[Desired],
+    current_key: Optional[Callable[[Current], Key]] = None,
+    desired_key: Optional[Callable[[Desired], Key]] = None,
+    equal: Optional[Callable[[Current, Desired], bool]] = None,
+) -> DiffResult[Current, Desired, Key]:
+    """
+    Compare two iterables and return a `DiffResult` instance containing the differences between them.
+
+    :param current: The current iterable to compare.
+    :type current: Iterable[Current]
+    :param desired: The desired iterable to compare.
+    :type desired: Iterable[Desired]
+    :param current_key: An optional function that returns the key for an element in the `current` iterable.
+        If not provided, the default behavior is to use the element itself as the key.
+    :type current_key: Optional[Callable[[Current], Key]]
+    :param desired_key: An optaionl function that returns the key for an element in the `desired` iterable.
+        If not provided, the default behavior is to use the element itself as the key.
+    :type desired_key: Optional[Callable[[Desired], Key]]
+    :param equal: An optional function that compares two elements in the iterables and returns True if they are equal.
+        If not provided, the default behavior is to use the `==` operator.
+    :type equal: Optional[Callable[[Current, Desired], bool]]
+    :return: A `DiffResult` instance containing the differences between the `current` and `desired` iterables,
+        including elements that were added, deleted, changed or identical.
+    :rtype: DiffResult[Current, Desired, Key]
+    :raises: None
+
+    Example:
+        >>> current = [
+        ...     {"name": "i", "value": 1},
+        ...     {"name": "c", "value": 2},
+        ...     {"name": "d", "value": 3},
+        ... ]
+        >>> desired = [
+        ...     {"name": "i", "value": 1},
+        ...     {"name": "c", "value": 20},
+        ...     {"name": "a", "value": 30},
+        ... ]
+        >>> result = diff_any_iterables(
+        ...     current,
+        ...     desired,
+        ...     lambda c: c["name"],
+        ...     lambda d: d["name"],
+        ...     equal=lambda c, d: c["value"] == d["value"],
+        ... )
+        DiffResult(
+            add={'a': {'name': 'a', 'value': 30}},
+            delete={'d': {'name': 'd', 'value': 3}},
+            change={'c': DiffPair(current={'name': 'c', 'value': 2}, desired={'name': 'c', 'value': 20})),
+            identical={'i': DiffPair(current={'name': 'i', 'value': 1}, desired={'name': 'i', 'value': 1})},
+        )
+    """
+    if current_key is None:
+        current_key = _default_current_key
+    if desired_key is None:
+        desired_key = _default_desired_key
```

Comment:
You can set both defaults in the method definition instead. 

```python
def diff_any_iterables(
     current: Iterable[Current],
     desired: Iterable[Desired],
     current_key: Optional[Callable[[Current], Key]] = _default_current_key,
     desired_key: Optional[Callable[[Desired], Key]] = _default_current_key,
     equal: Optional[Callable[[Current, Desired], bool]] = None,
 ) -> DiffResult[Current, Desired, Key]:
```

Reply:
Oh, thanks! I'm struggling making mypy happy, let me try this.

Reply:
It works!